### PR TITLE
Introducing the `Pipeline` replacement

### PR DESCRIPTION
**Pull request by WillJRoper**
This PR introduces the `Pipeline` class, a class for automating pipelines to reduce the amount of boilerplate a user needs to implement every time they want to write a pipeline for a new simulation or model.

## Issue Type

- Enhancement

## Checklist
- [x] I have read the [CONTRIBUTING.md]() -->
- [x] I have added docstrings to all methods
- [x] I have added sufficient comments to all lines
- [x] I have made corresponding changes to the documentation
- [x] My changes generate no pep8 errors
- [x] I have added tests that prove my fix is effective or that my feature works
- [x] New and existing unit tests pass locally with my changes

### CODE REVIEW

**Comment by WillJRoper** [comment]
For an out in the wild example see: https://github.com/WillJRoper/flares-lrds/blob/main/flares_survey.py

**Comment by WillJRoper** [comment]
This is actually ready now. Had missed some doc updates the first time I readied it.

**Comment by WillJRoper** [comment]
There are some changes outside the ``Pipeline`` introduction itself that were needed to make it work efficiently. In summary these are:

- Passing nthreads around where it was missing.
- Enabling I/O for saving instances of classes to HDF5.
- Allowing you to select a subset of a container.

**Comment by WillJRoper** [comment]
FYI, I want to add some options for limiting what is written out (probably flags passed to `write`) but I'll implement this in a future PR. The motivation being... with the images all saved my FLARES pipeline above is saving 7.2TB before doing z=5 😬

File: src/synthesizer/emission_models/base_model.py [hunk 1]
```diff
@@ -1434,6 +1438,62 @@ def fix_parameters(self, **kwargs):
...
+        """
+        Save the model to an HDF5 group.
+
+        Args:
+            group (h5py.Group):
+                The group to save the model to.
+        """
+        # First off call the operation to save operation specific attributes
+        # to the group
+        if self._is_extracting:
```
Comment (christopherlovell) [inline src/synthesizer/emission_models/base_model.py:1451]: I'm not sure why each of these models has to have a dedicated to_hdf5 method, can't an extraction / combination / etc. just have a general `to_hdf5` method, and then call it once here?

File: src/synthesizer/pipeline/pipeline.py [hunk 2]
```diff
@@ -0,0 +1,1801 @@
...
+            for f, img in imgs.items():
+                self.images_flux_psf["BlackHole"][spec_type][f] = unyt_array(
+                    img
+                )
+
+        # Done!
+        self._got_images_flux_psf = True
+        self._took(start, "Applying PSFs to flux images")
+
+    def get_lnu_data_cubes(self):
```
Comment (christopherlovell) [inline src/synthesizer/pipeline/pipeline.py:1407]: Why the4 different naming convention here compared to the spectra and images?

**Review by christopherlovell — CHANGES_REQUESTED** [review]

File: src/synthesizer/emission_models/base_model.py (hunk 1 above)
Comment (WillJRoper) [reply src/synthesizer/emission_models/base_model.py:1451]: It reduces the boilerplate to have a targeted writer for each operation. Each operation only has certain attributes populated. If you didn't isolate these then you'd have to have a check for each attr you're writing.

File: src/synthesizer/pipeline/pipeline.py (hunk 2 above)
Comment (WillJRoper) [reply src/synthesizer/pipeline/pipeline.py]: In short, no idea 😅 I had ignored them after adding the `NotImplemented` error.

File: src/synthesizer/emission_models/base_model.py (hunk 1 above)
Comment (WillJRoper) [reply src/synthesizer/emission_models/base_model.py]: Misunderstood, now fixed to use a single private `_to_hdf5` method on each operation

**Review by christopherlovell — APPROVED** [review]

**Pull request merged**